src/ucc/api/ucc.h

```diff
@@ -1574,5 +1588,229 @@ static inline ucc_status_t ucc_collective_test(ucc_coll_req_h request)
  */
 ucc_status_t ucc_collective_finalize(ucc_coll_req_h request);
 
+
+/**
+ * @ingroup UCC_EVENT_DT
+ *
+ */
+
+typedef enum ucc_event_type {
+	UCC_EVENT_COLLECTIVE_POST       = UCC_BIT(0),
+	UCC_EVENT_COLLECTIVE_COMPLETE   = UCC_BIT(1),
+	UCC_EVENT_COMPUTE_COMPLETE      = UCC_BIT(2),
+	UCC_EVENT_OVERFLOW              = UCC_BIT(3)
+} ucc_event_type_t;
+
+/**
+ * @ingroup UCC_EVENT_DT
+ *
+ */
+typedef enum ucc_ee_type {
+    UCC_CUDA_STREAM = UCC_BIT(0),
+	UCC_CPU_THREAD  = UCC_BIT(1),
+	UCC_DPU_THREAD  = UCC_BIT(2)
+} ucc_ee_type_t;
+
+/**
+ * @ingroup UCC_EVENT_DT
+ *
+ */
+typedef struct ucc_event_context {
+    ucc_event_type_t     ev_type;
+    size_t 		         info_size;
+    void 		         *info;
+} ucc_event_context_t;
+
+/**
+ * @ingroup UCC_EVENT_DT
+ *
+ */
+typedef struct ucc_event {
+	ucc_event_type_t        ev_type;
+	ucc_event_context_t     *event_context;
+	ucc_coll_req_h          req;
+} ucc_ev_t;
+
+/**
+ * @ingroup UCC_EVENT_DT
+ *
+ */
+typedef enum ucc_event_queue_type {
+	UCC_POST_QUEUE          = UCC_BIT(0),
+	UCC_COMPLETION_QUEUE    = UCC_BIT(1)
+} ucc_event_queue_type_t;
+
+/**
+ * @ingroup UCC_EVENT_DT
+ *
+ */
+typedef struct ucc_ee_params {
+     ucc_ee_type_t 			        ee_type;
+     ucc_event_queue_type_t        	ee_eq;
+     void 		     		        *ee_context;
+} ucc_ee_params_t;
+
+/**
+ * @ingroup UCC_EVENT_DT
+ *
+ */
+typedef struct ucc_ee_attribs {
+    ucc_event_queue_h   post_queue;
+    ucc_event_queue_h   completion_queue;
+} ucc_ee_attribs_t;
+
+/**
+ * @ingroup UCC_EVENT
+ *
+ * @brief The routine creates the execution context for collective operations.
+ *
+ * @param [in] team     team handle
+ * @param [in] params   user provided params to customize the execution engine
+ * @param [out] ee      execution engine handle
+ *
+ * @parblock
+ *
+ * @b Description
+ *
+ * @endparblock
+ *
+ * @return Error code as defined by ucc_status_t
+ */
+ ucc_status_t ucc_ee_create(ucc_team_h team, const ucc_ee_params_t *params, ucc_ee_h *ee);
+
+ /**
+ * @ingroup UCC_EVENT
+ *
+ * @brief The routine destroys the execution context created for collective operations.
+ *
+ * @param [in] ee   Execution engine handle
+ *
+ * @parblock
+ *
+ * @b Description
+ *
+ * @endparblock
+ *
+ * @return Error code as defined by ucc_status_t
+ */
+ ucc_status_t ucc_ee_destroy(ucc_ee_h ee);
+
+/**
+ * @ingroup UCC_EVENT
+ *
+ * @brief The routine gets the event from the event queue.
+ *
+ * @param [in]  ee        execution engine handle
+ * @param [in]  eq_type        Event queue to pull the event
+ * @param [out] ev        Event structure fetched from the event queue
+ *
+ * @parblock
+ *
+ * @b Description
+ *
+ * @endparblock
+ *
+ * @return Error code as defined by ucc_status_t
+ */
+ucc_status_t ucc_ee_get_event(ucc_ee_h ee, ucc_event_queue_type_t eq_type,
+                ucc_ev_t *ev);
```

Comment:
ev is a library object.  should we have ucc_event_destroy to release resources (memory..etc)

Reply:
May be delete/pop event would be appropriate? That way, library destroy's it.